File: frontend/src/utils/jsonLd.ts
/**
 * JSON-LD builders with strict whitelists.
 *
 * Spec: ПЛАН §6 Iter 4 — F9; БЗ SEO-индексация-SPA.md; security-baseline §3.
 */

import { catalogPathForSku } from "./catalogPaths";

const SITE_NAME = "Hoocon";
const SITE_URL = "https://hoocon.ru";

interface SkuForJsonLd {
  name: string;
  slug: string;
  sku_code?: string;
  description?: string;
  price?: string | number | null;
  price_on_request?: boolean;
  in_stock?: boolean;
  category_name?: string;
  category_slug?: string | null;
}

interface ArticleForJsonLd {
  title: string;
  slug: string;
  description: string;
  published_at?: string | null;
  pathPrefix?: "/statyi" | "/novosti";
}

interface BreadcrumbItem {
  name: string;
  path: string;
}

export type FaqJsonLdItem = { question: string; answer: string };

/**
 * Build a JSON-LD Product object from a SKU with a strict whitelist.
 */
export function buildProductJsonLd(sku: SkuForJsonLd): Record<string, unknown> {
  const path =
    sku.category_slug && sku.slug
      ? catalogPathForSku({
          category_slug: sku.category_slug,
          slug: sku.slug,
        })
      : `/${sku.slug}`;
  const ld: Record<string, unknown> = {
    "@context": "https://schema.org",
    "@type": "Product",
    name: sku.name,
    sku: sku.sku_code ?? sku.slug,
    url: `${SITE_URL}${path}`,
  };

  if (sku.description) {
    ld.description = sku.description;
  }
  if (sku.category_name) {
    ld.category = sku.category_name;
  }

  if (sku.price != null && !sku.price_on_request) {
    ld.offers = {
      "@type": "Offer",
      price: String(sku.price),
      priceCurrency: "RUB",
      availability:
        sku.in_stock === false
          ? "https://schema.org/PreOrder"
          : "https://schema.org/InStock",
    };
  } else {
    ld.offers = {
      "@type": "Offer",
      availability:
        sku.in_stock === false
          ? "https://schema.org/PreOrder"
          : "https://schema.org/InStock",
      priceSpecification: {
        "@type": "PriceSpecification",
        priceCurrency: "RUB",
      },
    };
  }

  return ld;
}

/** Organization + WebSite (+ SearchAction) for the home page. */
export function buildHomeJsonLd(
  faqItems: FaqJsonLdItem[] = [],
): Record<string, unknown>[] {
  const blocks: Record<string, unknown>[] = [
    {
      "@context": "https://schema.org",
      "@type": "Organization",
      name: SITE_NAME,
      url: SITE_URL,
    },
    {
      "@context": "https://schema.org",
      "@type": "WebSite",
      name: SITE_NAME,
      url: SITE_URL,
      potentialAction: {
        "@type": "SearchAction",
        target: {
          "@type": "EntryPoint",
          urlTemplate: `${SITE_URL}/search?q={search_term_string}`,
        },
        "query-input": "required name=search_term_string",
      },
    },
  ];
  if (faqItems.length) {
    blocks.push(buildFaqJsonLd(faqItems));
  }
  return blocks;
}

/** FAQPage schema from Admin FAQ items. */
export function buildFaqJsonLd(
  items: FaqJsonLdItem[],
): Record<string, unknown> {
  return {
    "@context": "https://schema.org",
    "@type": "FAQPage",
    mainEntity: items.map((item) => ({
      "@type": "Question",
      name: item.question,
      acceptedAnswer: { "@type": "Answer", text: item.answer },
    })),
  };
}

/** Article / BlogPosting for /statyi and /novosti. */
export function buildArticleJsonLd(
  article: ArticleForJsonLd,
): Record<string, unknown> {
  const prefix = article.pathPrefix ?? "/statyi";
  const ld: Record<string, unknown> = {
    "@context": "https://schema.org",
    "@type": "Article",
    headline: article.title,
    description: article.description,
    url: `${SITE_URL}${prefix}/${article.slug}`,
    author: { "@type": "Organization", name: SITE_NAME, url: SITE_URL },
    publisher: { "@type": "Organization", name: SITE_NAME, url: SITE_URL },
  };
  if (article.published_at) {
    ld.datePublished = article.published_at;
  }
  return ld;
}

/** BreadcrumbList from path segments. */
export function buildBreadcrumbJsonLd(
  items: BreadcrumbItem[],
): Record<string, unknown> {
  return {
    "@context": "https://schema.org",
    "@type": "BreadcrumbList",
    itemListElement: items.map((item, index) => ({
      "@type": "ListItem",
      position: index + 1,
      name: item.name,
      item: item.path.startsWith("http")
        ? item.path
        : `${SITE_URL}${item.path === "/" ? "" : item.path}`,
    })),
  };
}

File: backend/config/seo/json_ld.py
"""JSON-LD whitelist builders for source HTML (БЗ M2, Z3/Z4)."""

from __future__ import annotations

from typing import Any

from django.conf import settings

from config.seo.head import SeoHeadContext
from config.seo.routes import DEFAULT_DESCRIPTION, HOME_FAQ_ITEMS, SITE_NAME


def build_json_ld(context: SeoHeadContext) -> list[dict[str, Any]]:
    """Build schema.org blocks for a pathname (whitelist fields only).

    Args:
        context: Resolved SEO context.

    Returns:
        List of JSON-LD objects for injection into HTML.
    """
    site_url = settings.SITE_URL.rstrip("/")
    blocks: list[dict[str, Any]] = []

    if context.canonical_path == "/":
        blocks.append(_organization(site_url))
        blocks.append(_website(site_url))
        blocks.append(_faq_page(site_url, HOME_FAQ_ITEMS))
        return blocks

    if context.canonical_path == "/faq":
        blocks.append(_faq_page(site_url, HOME_FAQ_ITEMS))
        blocks.extend(_breadcrumb(site_url, context))
        return blocks

    if context.canonical_path == "/kontakty":
        blocks.append(_organization(site_url))
        blocks.extend(_breadcrumb(site_url, context))
        return blocks

    if context.sku_code and context.og_type == "product":
        blocks.append(_product(site_url, context))
        blocks.extend(_breadcrumb(site_url, context))
        return blocks

    if context.og_type == "article" and context.article_published_at:
        blocks.append(_article(site_url, context))
        blocks.extend(_breadcrumb(site_url, context))
        return blocks

    blocks.extend(_breadcrumb(site_url, context))
    return blocks


def _organization(site_url: str) -> dict[str, Any]:
    return {
        "@context": "https://schema.org",
        "@type": "Organization",
        "name": SITE_NAME,
        "url": site_url,
        "description": DEFAULT_DESCRIPTION,
    }


def _website(site_url: str) -> dict[str, Any]:
    return {
        "@context": "https://schema.org",
        "@type": "WebSite",
        "name": SITE_NAME,
        "url": site_url,
        "potentialAction": {
            "@type": "SearchAction",
            "target": {
                "@type": "EntryPoint",
                "urlTemplate": f"{site_url}/search?q={{search_term_string}}",
            },
            "query-input": "required name=search_term_string",
        },
    }


def _faq_page(site_url: str, items: tuple[tuple[str, str], ...]) -> dict[str, Any]:
    return {
        "@context": "https://schema.org",
        "@type": "FAQPage",
        "url": f"{site_url}/faq" if items else site_url,
        "mainEntity": [
            {
                "@type": "Question",
                "name": question,
                "acceptedAnswer": {"@type": "Answer", "text": answer},
            }
            for question, answer in items
        ],
    }


def _article(site_url: str, context: SeoHeadContext) -> dict[str, Any]:
    headline = context.page_title.split(" — ")[0]
    return {
        "@context": "https://schema.org",
        "@type": "Article",
        "headline": headline,
        "description": context.description,
        "url": f"{site_url}{context.canonical_path}",
        "datePublished": context.article_published_at,
        "author": {"@type": "Organization", "name": SITE_NAME, "url": site_url},
        "publisher": {"@type": "Organization", "name": SITE_NAME, "url": site_url},
    }


def _product(site_url: str, context: SeoHeadContext) -> dict[str, Any]:
    name = context.page_title.split(" — ")[0]
    # Storefront: zero stock = made-to-order (never «out of stock» UX).
    availability = "https://schema.org/InStock" if context.in_stock else "https://schema.org/PreOrder"
    offer: dict[str, Any] = {
        "@type": "Offer",
        "availability": availability,
    }
    if context.sku_price and not context.price_on_request:
        offer["price"] = context.sku_price
        offer["priceCurrency"] = "RUB"
    else:
        offer["priceSpecification"] = {
            "@type": "PriceSpecification",
            "priceCurrency": "RUB",
        }
    product: dict[str, Any] = {
        "@context": "https://schema.org",
        "@type": "Product",
        "name": name,
        "sku": context.sku_code,
        "description": context.description,
        "url": f"{site_url}{context.canonical_path}",
        "offers": offer,
    }
    if context.category_name:
        product["category"] = context.category_name
    return product


def _breadcrumb(site_url: str, context: SeoHeadContext) -> list[dict[str, Any]]:
    if len(context.breadcrumb) < 2:
        return []
    elements = []
    for index, (path, name) in enumerate(context.breadcrumb, start=1):
        item_url = f"{site_url}{path}" if path.startswith("/") else path
        elements.append(
            {
                "@type": "ListItem",
                "position": index,
                "name": name,
                "item": item_url,
            }
        )
    return [
        {
            "@context": "https://schema.org",
            "@type": "BreadcrumbList",
            "itemListElement": elements,
        }
    ]
